working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1646,7 +1771,7 @@ that contains the pattern:
 *   **Switch statement case**: The guard clause and the statements of the
     subsequent non-empty case body.
```

Comment:
If multiple `casehead`s can reach the same body, then we need some way to merge scopes.

Take:
```dart
 case Point(:int x, :int y):
 case (:int x, :int y):
    x += 1;
    y += 1;
    print(x + y);
```
Both branches introduce the same names with the same types. One of them reaches the body, and defines the variables that are modified. But *both* sets of variables can be in play:

```dart
late int Function() next;
...
 case Point(:int x, :int y) when (next = () => x++) != 0 && x < y:
 case Point(:int x, :int y): 
    print([x++, y++, next());
```
If given `Point(x:10, y: 5)` the first case will be rejected in the `when` clause, but still capture the `x` variable of that.
Then the next case succeeds, binding a new `x` and `y`.
It should print `[10, 5, 10]`.


Is this just an argument against allowing multiple `casehead`s for the same body? I think it is!

You can still capture variables from *other* cases, which were rejected in the `when` clauses, but those will definitely be *different* variables, but you only get one pattern scope and one `when` clause per body.

```dart
late int Function() next;
...
 case Point(:int x, :int y) when (next = () => x++) != 0 && false:
    print("not today");
 case Point(:int x, :int y): 
    print([x++, y++, next());
```
This still works and binds the variable of the first (matching!) pattern, but there is no way for those variables to ever flow into the same body as the second `case`.

We can allow `case c1: case c2: case c3: body` as a backwards compatible way to write `case c1 | c2 | c3: body`, but you can't add bindings or `when` clauses to any of those cases. To do that, you have to combine them to one pattern using `|`.

On the other hand, that would prohibit:
```dart
case Value(:var x) when x >= 0:
case NegatedValue(:var x) when x < 0:
  same body;
```
if that's something you want to do (multiple independent `when`s leading to the same body).